> @ibuclaw maybe you should do cstuff2.c to avoid conflicting additions!

Done in #12597.

I've just ran the importC testsuite I [knocked together](https://github.com/ibuclaw/importC), and I get further now, so at least one of the listed issues are fixed by this PR.  Good!
```
../dmd/generated/linux/release/64/dmd -c generated/stdc/assert_.c -ofgenerated/stdc/assert_.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/complex.c -ofgenerated/stdc/complex.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/ctype.c -ofgenerated/stdc/ctype.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/errno.c -ofgenerated/stdc/errno.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/fenv.c -ofgenerated/stdc/fenv.o
generated/stdc/fenv.c(37): Error: bit fields are not supported
generated/stdc/fenv.c(38): Error: bit fields are not supported
make: *** [Makefile:23: generated/stdc/fenv.o] Error 1
../dmd/generated/linux/release/64/dmd -c generated/stdc/float_.c -ofgenerated/stdc/float_.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/inttypes.c -ofgenerated/stdc/inttypes.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/limits.c -ofgenerated/stdc/limits.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/locale.c -ofgenerated/stdc/locale.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/math.c -ofgenerated/stdc/math.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/signal.c -ofgenerated/stdc/signal.o
generated/stdc/signal.c(284): Error: function `signal.sigaction` conflicts with struct `signal.sigaction` at generated/stdc/signal.c(268)
generated/stdc/signal.c(450): Error: function `signal.sigstack` conflicts with struct `signal.sigstack` at generated/stdc/signal.c(444)
make: *** [Makefile:23: generated/stdc/signal.o] Error 1
../dmd/generated/linux/release/64/dmd -c generated/stdc/stdarg.c -ofgenerated/stdc/stdarg.o
Error: undefined identifier `__builtin_va_list`
make: *** [Makefile:23: generated/stdc/stdarg.o] Error 1
../dmd/generated/linux/release/64/dmd -c generated/stdc/stddef.c -ofgenerated/stdc/stddef.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/stdint.c -ofgenerated/stdc/stdint.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/stdio.c -ofgenerated/stdc/stdio.o
generated/stdc/stdio.c(88): Error: struct `stdio._IO_FILE` conflicts with struct `stdio._IO_FILE` at generated/stdc/stdio.c(86)
generated/stdc/stdio.c(90): Error: struct `stdio._IO_FILE` conflicts with struct `stdio._IO_FILE` at generated/stdc/stdio.c(86)
generated/stdc/stdio.c(95): Error: struct `stdio._IO_FILE` conflicts with struct `stdio._IO_FILE` at generated/stdc/stdio.c(86)
make: *** [Makefile:23: generated/stdc/stdio.o] Error 1
../dmd/generated/linux/release/64/dmd -c generated/stdc/stdlib.c -ofgenerated/stdc/stdlib.o
generated/stdc/stdlib.c(320): Error: alias `stdlib.pthread_attr_t` conflicts with union `stdlib.pthread_attr_t` at generated/stdc/stdlib.c(315)
make: *** [Makefile:23: generated/stdc/stdlib.o] Error 1
../dmd/generated/linux/release/64/dmd -c generated/stdc/string.c -ofgenerated/stdc/string.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/tgmath.c -ofgenerated/stdc/tgmath.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/time.c -ofgenerated/stdc/time.o
../dmd/generated/linux/release/64/dmd -c generated/stdc/wchar_.c -ofgenerated/stdc/wchar_.o
generated/stdc/wchar_.c(17): Error: struct `wchar_._IO_FILE` conflicts with struct `wchar_._IO_FILE` at generated/stdc/wchar_.c(15)
make: *** [Makefile:23: generated/stdc/wchar_.o] Error 1
../dmd/generated/linux/release/64/dmd -c generated/stdc/wctype.c -ofgenerated/stdc/wctype.o
generated/stdc/wctype.c(80): Error: expression expected, not `<`
generated/stdc/wctype.c(80): Error: found `8` when expecting `)`
generated/stdc/wctype.c(80): Error: expression expected, not `)`
generated/stdc/wctype.c(80): Error: found `:` when expecting `)`
generated/stdc/wctype.c(80): Error: expression expected, not `<`
generated/stdc/wctype.c(80): Error: found `16` when expecting `)`
generated/stdc/wctype.c(80): Error: expression expected, not `)`
generated/stdc/wctype.c(80): Error: found `:` when expecting `)`
generated/stdc/wctype.c(80): Error: expression expected, not `<`
generated/stdc/wctype.c(80): Error: found `24` when expecting `)`
generated/stdc/wctype.c(80): Error: expression expected, not `)`
generated/stdc/wctype.c(80): Error: found `:` when expecting `)`
generated/stdc/wctype.c(80): Error: expression expected, not `int`
generated/stdc/wctype.c(80): Error: expression expected, not `)`
generated/stdc/wctype.c(80): Error: found `)` when expecting `:`
generated/stdc/wctype.c(80): Error: expression expected, not `)`
generated/stdc/wctype.c(81): Error: expression expected, not `<`
generated/stdc/wctype.c(81): Error: found `8` when expecting `)`
generated/stdc/wctype.c(81): Error: expression expected, not `)`
generated/stdc/wctype.c(81): Error: found `:` when expecting `)`
make: *** [Makefile:23: generated/stdc/wctype.o] Error 1
make: Target 'all' not remade because of errors.
```

--- a/src/dmd/astbase.d
+++ b/src/dmd/astbase.d
@@ -206,6 +206,7 @@ struct ASTBase
         Ttraits,
         Tmixin,
         Tnoreturn,
+        Ttag,
         TMAX
     }
 
@@ -256,6 +257,7 @@ struct ASTBase
     alias Ttraits = ENUMTY.Ttraits;
     alias Tmixin = ENUMTY.Tmixin;
     alias Tnoreturn = ENUMTY.Tnoreturn;
+    alias Ttag = ENUMTY.Ttag;
     alias TMAX = ENUMTY.TMAX;
 
     alias TY = ubyte;
@@ -2823,6 +2825,7 @@ struct ASTBase
                 sizeTy[Tvector] = __traits(classInstanceSize, TypeVector);
                 sizeTy[Tmixin] = __traits(classInstanceSize, TypeMixin);
                 sizeTy[Tnoreturn] = __traits(classInstanceSize, TypeNoreturn);
+                sizeTy[Ttag] = __traits(classInstanceSize, TypeTag);
                 return sizeTy;
             }();
 
@@ -3588,6 +3591,7 @@ struct ASTBase
             inout(TypeNull)       isTypeNull()       { return ty == Tnull      ? cast(typeof(return))this : null; }
             inout(TypeMixin)      isTypeMixin()      { return ty == Tmixin     ? cast(typeof(return))this : null; }
             inout(TypeTraits)     isTypeTraits()     { return ty == Ttraits    ? cast(typeof(return))this : null; }
+            inout(TypeTag)        isTypeTag()        { return ty == Ttag       ? cast(typeof(return))this : null; }
         }
 
         override void accept(Visitor v)
@@ -3924,6 +3928,36 @@ struct ASTBase
         }
 
         override TypeStruct syntaxCopy()
+        {
+            return this;
+        }
+
+        override void accept(Visitor v)
+        {
+            v.visit(this);
+        }
+    }
+
+    extern (C++) final class TypeTag : Type
+    {
+        Loc loc;
+        TOK tok;
+        Identifier id;
+        Dsymbols* members;
+
+        Type resolved;
+
+        extern (D) this(const ref Loc loc, TOK tok, Identifier id, Dsymbols* members)
+        {
+            //printf("TypeTag %p\n", this);
+            super(Ttag);
+            this.loc = loc;
+            this.tok = tok;
+            this.id = id;
+            this.members = members;
+        }
+
+        override TypeTag syntaxCopy()
         {
             return this;
         }

--- a/src/dmd/cparse.d
+++ b/src/dmd/cparse.d
@@ -2798,9 +2798,10 @@ final class CParser(AST) : Parser!AST
      * Returns:
      *  type of the struct
      */
-    private AST.TypeStruct cparseStruct(ref AST.Dsymbols* symbols)
+    private AST.Type cparseStruct(ref AST.Dsymbols* symbols)
     {
-        auto structOrUnion = token;
+        const structOrUnion = token.value;
+        assert(structOrUnion == TOK.struct_ || structOrUnion == TOK.union_);
         const loc = token.loc;
         nextToken();
 
@@ -2813,23 +2814,39 @@ final class CParser(AST) : Parser!AST
             cparseGnuAttributes();
 
         Identifier tag;
+
+        // declare `tag` as symbol
+        AST.Type declareTag()
+        {
+            auto stag = (structOrUnion == TOK.struct_)
+                ? new AST.StructDeclaration(loc, tag, false)
+                : new AST.UnionDeclaration(loc, tag);
+            if (!symbols)
+                symbols = new AST.Dsymbols();
+            symbols.push(stag);
+            return new AST.TypeStruct(stag);
+        }
+
         if (token.value == TOK.identifier)
         {
             tag = token.ident;
             nextToken();
+            if (token.value == TOK.semicolon)
+            {
+                /* `struct tag;` always results in a declaration
+                 * in the current scope
+                 */
+                return declareTag();
+            }
         }
 
-        assert(structOrUnion.value == TOK.struct_ || structOrUnion.value == TOK.union_);
-        auto stag = (structOrUnion.value == TOK.struct_)
-            ? new AST.StructDeclaration(loc, tag, false)
-            : new AST.UnionDeclaration(loc, tag);
-
-        if (!symbols)
-            symbols = new AST.Dsymbols();
-        symbols.push(stag);
-
+        AST.Dsymbols* members;
         if (token.value == TOK.leftCurly)
         {
+            // Give it a name if it doesn't have one
+            if (!tag)
+                tag = Identifier.generateId("__tag"[]);
+
             nextToken();
             auto symbolsSave = symbols;
             symbols = null;
@@ -2840,7 +2857,7 @@ final class CParser(AST) : Parser!AST
                 if (token.value == TOK.endOfFile)
                     break;
             }
-            stag.members = symbols;
+            members = symbols;
             symbols = symbolsSave;
             check(TOK.rightCurly);
 
@@ -2850,13 +2867,24 @@ final class CParser(AST) : Parser!AST
             if (token.value == TOK.__attribute__)
                 cparseGnuAttributes();
 
-            if (tag && (!stag.members || stag.members.length == 0)) // C11 6.7.2.1-2
-                error("no struct-declarator-list for `%s %s`", structOrUnion.toChars(), tag.toChars());
+            if (tag && (!members || (*members).length == 0)) // C11 6.7.2.1-2
+                error("no struct-declarator-list for `%s %s`", Token.toChars(structOrUnion), tag.toChars());
         }
         else if (!tag)
-            error("missing `identifier` after `%s`", structOrUnion.toChars());
+            error("missing `identifier` after `%s`", Token.toChars(structOrUnion));
 
-        return new AST.TypeStruct(stag);
+        if (token.value == TOK.semicolon)
+        {
+            /* `struct tag { ... };` always results in a declaration
+             * in the current scope
+             */
+            return declareTag();
+        }
+        /* Need semantic information to determine if this is a declaration,
+         * redeclaration, or reference to existing declaration.
+         * Defer to the semantic() pass with a TypeTag.
+         */
+        return new AST.TypeTag(loc, structOrUnion, tag, members);
     }
 
     //}

--- a/src/dmd/dmangle.d
+++ b/src/dmd/dmangle.d
@@ -236,6 +236,7 @@ private immutable char[TMAX] mangleChar =
     Tvector      : '@',
     Ttraits      : '@',
     Tmixin       : '@',
+    Ttag         : '@',
     Tnoreturn    : '@',         // becomes 'Nn'
 ];
 

--- a/src/dmd/dsymbolsem.d
+++ b/src/dmd/dsymbolsem.d
@@ -4290,7 +4290,7 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
             return;
         int errors = global.errors;
 
-        //printf("+StructDeclaration::semantic(this=%p, '%s', sizeok = %d)\n", this, toPrettyChars(), sizeok);
+        //printf("+StructDeclaration::semantic(this=%p, '%s', sizeok = %d)\n", this, sd.toPrettyChars(), sd.sizeok);
         Scope* scx = null;
         if (sd._scope)
         {

--- a/src/dmd/frontend.h
+++ b/src/dmd/frontend.h
@@ -707,7 +707,8 @@ enum class TY : uint8_t
     Ttraits = 44u,
     Tmixin = 45u,
     Tnoreturn = 46u,
-    TMAX = 47u,
+    Ttag = 47u,
+    TMAX = 48u,
 };
 
 struct Mcache;
@@ -733,6 +734,7 @@ class TypeNull;
 class TypeMixin;
 class TypeTraits;
 class TypeNoreturn;
+class TypeTag;
 class TemplateTypeParameter;
 class TemplateValueParameter;
 class TemplateAliasParameter;
@@ -1355,7 +1357,7 @@ public:
     static ClassDeclaration* typeinfoshared;
     static ClassDeclaration* typeinfowild;
     static TemplateDeclaration* rtinfo;
-    static Type* basic[47LLU];
+    static Type* basic[48LLU];
     virtual const char* kind() const;
     Type* copy() const;
     virtual Type* syntaxCopy();
@@ -1475,6 +1477,7 @@ public:
     TypeMixin* isTypeMixin();
     TypeTraits* isTypeTraits();
     TypeNoreturn* isTypeNoreturn();
+    TypeTag* isTypeTag();
     void accept(Visitor* v);
     TypeFunction* toTypeFunction();
 };
@@ -5514,6 +5517,19 @@ public:
     void accept(Visitor* v);
 };
 
+class TypeTag final : public Type
+{
+public:
+    Loc loc;
+    TOK tok;
+    Identifier* id;
+    Array<Dsymbol* >* members;
+    Type* resolved;
+    const char* kind() const;
+    TypeTag* syntaxCopy();
+    void accept(Visitor* v);
+};
+
 class Parameter final : public ASTNode
 {
 public:
@@ -5771,6 +5787,7 @@ public:
     virtual void visit(typename AST::TypeError t);
     virtual void visit(typename AST::TypeNull t);
     virtual void visit(typename AST::TypeNoreturn t);
+    virtual void visit(typename AST::TypeTag t);
     virtual void visit(typename AST::TypeVector t);
     virtual void visit(typename AST::TypeEnum t);
     virtual void visit(typename AST::TypeTuple t);

--- a/src/dmd/hdrgen.d
+++ b/src/dmd/hdrgen.d
@@ -3802,6 +3802,14 @@ private void typeToBufferx(Type t, OutBuffer* buf, HdrGenState* hgs)
             buf.writestring(hgs.fullQual ? t.sym.toPrettyChars() : t.sym.toChars());
     }
 
+    void visitTag(TypeTag t)
+    {
+        buf.writestring(Token.toChars(t.tok));
+        buf.writeByte(' ');
+        if (t.id)
+            buf.writestring(t.id.toChars());
+    }
+
     void visitTuple(TypeTuple t)
     {
         parametersToBuffer(ParameterList(t.arguments, VarArg.none), buf, hgs);
@@ -3863,5 +3871,6 @@ private void typeToBufferx(Type t, OutBuffer* buf, HdrGenState* hgs)
         case Tnull:      return visitNull(cast(TypeNull)t);
         case Tmixin:     return visitMixin(cast(TypeMixin)t);
         case Tnoreturn:  return visitNoreturn(cast(TypeNoreturn)t);
+        case Ttag:       return visitTag(cast(TypeTag)t);
     }
 }

--- a/src/dmd/mtype.d
+++ b/src/dmd/mtype.d
@@ -286,6 +286,7 @@ enum TY : ubyte
     Ttraits,
     Tmixin,
     Tnoreturn,
+    Ttag,
     TMAX,
 }
 
@@ -336,6 +337,7 @@ alias Tuns128 = ENUMTY.Tuns128;
 alias Ttraits = ENUMTY.Ttraits;
 alias Tmixin = ENUMTY.Tmixin;
 alias Tnoreturn = ENUMTY.Tnoreturn;
+alias Ttag = ENUMTY.Ttag;
 alias TMAX = ENUMTY.TMAX;
 
 alias ENUMTY = TY;
@@ -505,6 +507,7 @@ extern (C++) abstract class Type : ASTNode
             sizeTy[Ttraits] = __traits(classInstanceSize, TypeTraits);
             sizeTy[Tmixin] = __traits(classInstanceSize, TypeMixin);
             sizeTy[Tnoreturn] = __traits(classInstanceSize, TypeNoreturn);
+            sizeTy[Ttag] = __traits(classInstanceSize, TypeTag);
             return sizeTy;
         }();
 
@@ -2725,6 +2728,7 @@ extern (C++) abstract class Type : ASTNode
         inout(TypeMixin)      isTypeMixin()      { return ty == Tmixin     ? cast(typeof(return))this : null; }
         inout(TypeTraits)     isTypeTraits()     { return ty == Ttraits    ? cast(typeof(return))this : null; }
         inout(TypeNoreturn)   isTypeNoreturn()   { return ty == Tnoreturn  ? cast(typeof(return))this : null; }
+        inout(TypeTag)        isTypeTag()        { return ty == Ttag       ? cast(typeof(return))this : null; }
     }
 
     override void accept(Visitor v)
@@ -6769,6 +6773,52 @@ extern (C++) final class TypeNoreturn : Type
     override uint alignsize()
     {
         return 0;
+    }
+
+    override void accept(Visitor v)
+    {
+        v.visit(this);
+    }
+}
+
+/***********************************************************
+ * Unlike D, C can declare/define struct/union/enum tag names
+ * inside Declarators, instead of separately as in D.
+ * The order these appear in the symbol table must be in lexical
+ * order. There isn't enough info at the parsing stage to determine if
+ * it's a declaration or a reference to an existing name, so this Type
+ * collects the necessary info and defers it to semantic().
+ */
+extern (C++) final class TypeTag : Type
+{
+    Loc loc;                /// location of declaration
+    TOK tok;                /// TOK.struct_, TOK.union_, TOK.enum_
+    Identifier id;          /// tag name identifier
+    Dsymbols* members;      /// members of struct, null if none
+
+    Type resolved;          /// type after semantic() in case there are more others
+                            /// pointing to this instance, which can happen with
+                            ///   struct S { int a; } s1, *s2;
+
+    extern (D) this(const ref Loc loc, TOK tok, Identifier id, Dsymbols* members)
+    {
+        //printf("TypeTag %p\n", this);
+        super(Ttag);
+        this.loc = loc;
+        this.tok = tok;
+        this.id = id;
+        this.members = members;
+    }
+
+    override const(char)* kind() const
+    {
+        return "tag";
+    }
+
+    override TypeTag syntaxCopy()
+    {
+        // No semantic analysis done, no need to copy
+        return this;
     }
 
     override void accept(Visitor v)

--- a/src/dmd/mtype.h
+++ b/src/dmd/mtype.h
@@ -335,6 +335,7 @@ public:
     TypeMixin *isTypeMixin();
     TypeTraits *isTypeTraits();
     TypeNoreturn *isTypeNoreturn();
+    TypeTag *isTypeTag();
 
     void accept(Visitor *v) { v->visit(this); }
 };
@@ -897,6 +898,14 @@ public:
     bool isBoolean() /* const */;
     d_uns64 size(const Loc& loc) /* const */;
     unsigned alignsize();
+
+    void accept(Visitor *v) { v->visit(this); }
+};
+
+class TypeTag final : public Type
+{
+public:
+    TypeTag *syntaxCopy();
 
     void accept(Visitor *v) { v->visit(this); }
 };

--- a/src/dmd/parsetimevisitor.d
+++ b/src/dmd/parsetimevisitor.d
@@ -141,6 +141,7 @@ public:
     void visit(AST.TypeError t) { visit(cast(AST.Type)t); }
     void visit(AST.TypeNull t) { visit(cast(AST.Type)t); }
     void visit(AST.TypeNoreturn t) { visit(cast(AST.Type)t); }
+    void visit(AST.TypeTag t) { visit(cast(AST.Type)t); }
     void visit(AST.TypeVector t) { visit(cast(AST.Type)t); }
     void visit(AST.TypeEnum t) { visit(cast(AST.Type)t); }
     void visit(AST.TypeTuple t) { visit(cast(AST.Type)t); }

--- a/src/dmd/typesem.d
+++ b/src/dmd/typesem.d
@@ -1991,6 +1991,147 @@ extern(C++) Type typeSemantic(Type type, const ref Loc loc, Scope* sc)
         return error();
     }
 
+    Type visitTag(TypeTag mtype)
+    {
+        //printf("TypeTag.semantic() %s\n", mtype.toChars());
+        if (mtype.resolved)
+        {
+            /* struct S s, *p;
+             */
+            //printf("already resolved\n");
+            return mtype.resolved;
+        }
+
+        // TODO: Refactor this function into dscope.d, as
+        // well as other uses of this logic in dmd
+        static Scope* inner(Scope* sc)
+        {
+            for (auto scx = sc; ; scx = scx.enclosing)
+            {
+                if (scx.scopesym)
+                    return scx;
+            }
+        }
+
+        /* Declare mtype as a struct/union/enum declaration
+         */
+        void declareTag()
+        {
+            if (mtype.tok == TOK.enum_)
+            {
+                .error(mtype.loc, "enum not implemented");
+                mtype.resolved = error();
+            }
+            else
+            {
+                auto sd = new StructDeclaration(mtype.loc, mtype.id, false);
+                sd.members = mtype.members;
+                auto scopesym = inner(sc).scopesym;
+                if (scopesym.members)
+                    scopesym.members.push(sd);
+                sd.parent = sc.parent;
+                sd.dsymbolSemantic(sc);
+                mtype.resolved = visitStruct(new TypeStruct(sd));
+            }
+        }
+
+        /* look for pre-existing declaration
+         * TODO: tag namespace not implemented yet
+         * TODO: semantics for enums are slightly different, not implemented yet
+         */
+        Dsymbol scopesym;
+        auto s = sc.search(mtype.loc, mtype.id, &scopesym, IgnoreErrors);
+        if (!s)
+        {
+            // no pre-existing declaration, so declare it
+            declareTag();
+            return mtype.resolved;
+        }
+
+        /* A redeclaration only happens if both declarations are in
+         * the same scope
+         */
+        const bool redeclar = (scopesym == inner(sc).scopesym);
+
+        if (redeclar)
+        {
+            if (mtype.tok == TOK.enum_ && s.isEnumDeclaration())
+            {
+                auto ed = s.isEnumDeclaration();
+                if (mtype.members && ed.members)
+                    .error(mtype.loc, "`%s` already has members", mtype.id.toChars());
+                else
+                    ed.members = mtype.members;
+                mtype.resolved = ed.type;
+            }
+            else if (mtype.tok == TOK.union_ && s.isUnionDeclaration() ||
+                     mtype.tok == TOK.struct_ && s.isStructDeclaration())
+            {
+                // Add members to original declaration
+                auto sd = s.isStructDeclaration();
+                if (mtype.members && sd.members)
+                {
+                    /* struct S { int b; };
+                     * struct S { int a; } *s;
+                     */
+                    .error(mtype.loc, "`%s` already has members", mtype.id.toChars());
+                }
+                else if (!sd.members)
+                {
+                    /* struct S;
+                     * struct S { int a; } *s;
+                     */
+                    sd.members = mtype.members;
+                }
+                else
+                {
+                    /* struct S { int a; };
+                     * struct S *s;
+                     */
+                }
+                mtype.resolved = sd.type;
+            }
+            else
+            {
+                /* int S;
+                 * struct S { int a; } *s;
+                 */
+                .error(mtype.loc, "redeclaration of `%s`", mtype.id.toChars());
+                mtype.resolved = error();
+            }
+        }
+        else if (mtype.members)
+        {
+            /* struct S;
+             * { struct S { int a; } *s; }
+             */
+            declareTag();
+        }
+        else
+        {
+            if (mtype.tok == TOK.enum_ && s.isEnumDeclaration())
+            {
+                mtype.resolved = s.isEnumDeclaration().type;
+            }
+            else if (mtype.tok == TOK.union_ && s.isUnionDeclaration() ||
+                     mtype.tok == TOK.struct_ && s.isStructDeclaration())
+            {
+                /* struct S;
+                 * { struct S *s; }
+                 */
+                mtype.resolved = s.isStructDeclaration().type;
+            }
+            else
+            {
+                /* int S;
+                 * { struct S *s; }
+                 */
+                declareTag();
+            }
+        }
+        return mtype.resolved;
+    }
+
     switch (type.ty)
     {
         default:         return visitType(type);
@@ -2013,6 +2154,7 @@ extern(C++) Type typeSemantic(Type type, const ref Loc loc, Scope* sc)
         case Ttuple:     return visitTuple (cast(TypeTuple)type);
         case Tslice:     return visitSlice(cast(TypeSlice)type);
         case Tmixin:     return visitMixin(cast(TypeMixin)type);
+        case Ttag:       return visitTag(cast(TypeTag)type);
     }
 }
 

--- a/src/dmd/visitor.h
+++ b/src/dmd/visitor.h
@@ -84,6 +84,7 @@ class TypeNull;
 class TypeNoreturn;
 class TypeTraits;
 class TypeMixin;
+class TypeTag;
 
 class Dsymbol;
 
@@ -437,6 +438,7 @@ public:
     virtual void visit(TypeQualified *t) { visit((Type *)t); }
     virtual void visit(TypeTraits *t) { visit((Type *)t); }
     virtual void visit(TypeMixin *t) { visit((Type *)t); }
+    virtual void visit(TypeTag *t) { visit((Type *)t); }
 
     // TypeNext
     virtual void visit(TypeReference *t) { visit((TypeNext *)t); }

--- a/test/compilable/imports/cstuff1.c
+++ b/test/compilable/imports/cstuff1.c
@@ -184,6 +184,13 @@ void tokens()
 
 /********************************/
 
+void test__func__()
+{
+    _Static_assert((sizeof __func__) == 13, "ok");
+}
+
+/********************************/
+
 struct SA { int a, b, *const c, d[50]; };
 
 int test1(int i)
@@ -289,14 +296,38 @@ void test4(int i)
     long long ll = 12L;
     ll = 12UL;
     long double ld = 1.0L;
-    char* s = "hello" "betty";
+    const char* s = "hello" "betty";
 }
 
 /********************************/
 
-void test__func__()
+struct SS { int a; };
+
+int tests1()
 {
-    _Static_assert((sizeof __func__) == 13, "ok");
+    struct SS s;
+    return s.a;
+}
+
+int tests2()
+{
+    struct T { int a; };
+    struct T t;
+    return t.a;
+}
+
+void* tests3()
+{
+    struct T1 *p;
+    return p;
+}
+
+int tests4()
+{
+    struct S { int b; } a, *p;
+    a.b = 3;
+    p = &a;
+    return p->b;
 }
 
 /********************************/

